hoodie-client/src/main/java/com/uber/hoodie/table/HoodieMergeOnReadTable.java

```diff
@@ -371,6 +373,119 @@ public Partitioner getUpsertPartitioner(WorkloadProfile profile) {
           .map(smallFile -> ((SmallFile) smallFile).location.getFileId())
           .collect(Collectors.toList());
     }
+
+    private long getTotalFileSize(String partitionPath, FileSlice fileSlice) {
+      if (rollingStatMetadata != null) {
+        Map<String, HoodieRollingStat> partitionRollingStats =
+            rollingStatMetadata.getPartitionToRollingStats().get(partitionPath);
+        if (partitionRollingStats != null) {
+          HoodieRollingStat rollingStatForFile = partitionRollingStats.get(fileSlice.getFileId());
+          if (rollingStatForFile != null) {
+            long inserts = rollingStatForFile.getInserts();
+            long totalSize = averageRecordSize * inserts;
+            return totalSize;
+          }
+        }
+      }
+      // In case Rolling Stats is not present, fall back to sizing log files based on heuristics
+      if (!fileSlice.getDataFile().isPresent()) {
+        return convertLogFilesSizeToExpectedParquetSize(fileSlice.getLogFiles().collect(Collectors.toList()));
+      } else {
+        return fileSlice.getDataFile().get().getFileSize() + convertLogFilesSizeToExpectedParquetSize(fileSlice
+            .getLogFiles().collect(Collectors.toList()));
+      }
+    }
+
+    private boolean isSmallFile(String partitionPath, FileSlice fileSlice) {
+      long totalSize = getTotalFileSize(partitionPath, fileSlice);
+      if (totalSize < config.getParquetMaxFileSize()) {
+        return true;
+      }
+      return false;
+    }
+
+    // TODO (NA) : Make this static part of utility
+    protected long convertLogFilesSizeToExpectedParquetSize(List<HoodieLogFile> hoodieLogFiles) {
```

Comment:
add some tests for this?

Reply:
I've added some extra cases to testUpsertPartitioner already which should cover this but I agree we should have more granular unit test here. Let me think about this.